### DISCUSSION TITLE
Why don't we use *Pytest* in our tests?

### DISCUSSION THREAD

**Discussion by saroad2** [beeware/toga#1598] [Ideas]
So this question was bugging me for some time.

As for the moment of writing this post, *Toga* uses *Pytest* as the tests runner via *Tox*, but it does not use it as a library inside the actual tests. instead, we only use the default *unittest* library.

*Pytest* has a lot to offer as a testing framework:
1. Fixtures that can be used in multiple tests and reduce a lot of lines of code
2. A big collection of cool plugins that can help us with our testing. My favorite plugin is *pytest-cases* which can be used to create Data-Driven-Tests (DDT) in a very easy way.
3. easy ways of parametrizing, skipping and debugging
4. A vast community that keep improving the *Pytest* library
5. etc.

*Pytest* is definately not the only python testing framework out there. There are *nose*, *testify*, *lattuce* and others.
I think that of all the possible options, keeping with *unittest* is my least favorite. Even though it comes out-of-the-box, it's a little bit outdated in todays standards.

If we do choose to adopt a testing framework, the change does not have to happen immidiately. We can start slowly convert new tests to the framework when we get there.

Those are my thoughts. I would love to hear what you think.

**Comment by freakboy3742** [comment]
Completely agreed that a move to pytest is warranted.

We're already using pytest to *run* the test suite. The only reason that we aren't using pytest *tests* is legacy. When I started Toga, I wasn't a pytest convert, and as a result, there's a lot of older tests (and, more importantly, test *infrastructure*) that uses unittest. 

Modernising the test suite (and cleaning out a lot of the cruft of the test suite) is one of the housekeeping/foundation pieces of work that I'm hoping to tackle over the coming months. That will likely involve migrating to pytest in some fashion.
